spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/query/AbstractSpannerQuery.java

```diff
@@ -0,0 +1,84 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.repository.query;
+
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import org.springframework.cloud.gcp.data.spanner.core.SpannerOperations;
+import org.springframework.cloud.gcp.data.spanner.core.convert.ConversionUtils;
+import org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerMappingContext;
+import org.springframework.data.repository.query.QueryMethod;
+import org.springframework.data.repository.query.RepositoryQuery;
+
+/**
+ * @author Chengyuan Zhao
+ */
+abstract class AbstractSpannerQuery implements RepositoryQuery {
+
+	protected final QueryMethod queryMethod;
+
+	protected final SpannerOperations spannerOperations;
+
+	protected final SpannerMappingContext spannerMappingContext;
+
+	protected final Class entityType;
+
+	/**
+	 * Constructor
+	 * @param type the underlying entity type
+	 * @param queryMethod the underlying query method to support.
+	 * @param spannerOperations used for executing queries.
+	 * @param spannerMappingContext used for getting metadata about entities.
+	 */
+	AbstractSpannerQuery(Class type, QueryMethod queryMethod,
+			SpannerOperations spannerOperations,
+			SpannerMappingContext spannerMappingContext) {
+		this.queryMethod = queryMethod;
+		this.entityType = type;
+		this.spannerOperations = spannerOperations;
+		this.spannerMappingContext = spannerMappingContext;
+	}
+
+	@Override
+	public Object execute(Object[] parameters) {
+		Object rawResult = executeRawResult(parameters);
```

Comment:
Can you please either show an example that this is how in other Spring Data modules is done? 
My worry is that creating the full parent entity can cause surprises/unwanted performance impact.

Reply:
Take a look at `AbstractJpaQuery.java` from the spring-data-jpa project. They do it in a nearly identical way.

Regardless, I think it would be a bad idea at this point to carry the projection information any higher up in the query/retrieval logic. Furthermore, while it is very very user-friendly to create a projection using spring-data  provided query method and result processor, they do not let you extract the needed projection requirements one  would need to actually that info upward to the query/retrieval code. 

Also, I think the assumption you're coming with that there might be some performance-limited situation where a projection-returning method works well but the actual domain type of the repository in which the projection-returning method lives would not is really not what I'm getting reading the projection docs from Spring Data. I'm really surprised there is an interpretation of the purpose of projections function as anything other than syntax-related convenience.

Reply:
Thank you, I'm convinced that this solution is in line with the Spring Data design. 

I'm still not in love with it though :) - it's a hardcoded depedency between the entity creation and the project creation and doesn't allow you to be more surgical about it. But it's fine - we might want to mention though this in the user docs that user can define their subselect if they want to and relax the invariants on the entity for partial reading.
Let's add a section to the ref-doc to explain this?

Reply:
Will do ! 

Though one thing we is we made the decision early on the throw an exception if a column in the POJO doesn't appear in the resultset from Spanner. We originally did this to ensure that POJOs were coming from their home table and to prevent people playing fast and loose. Since we're still using the same read converter code, if the user selects too few columns from Spanner in their query then they will get that error, because it is still the full POJO in the back. 

we could just remove that restriction. what do you think?

Reply:
Oh i have an idea. By default we'll throw the exception but in the SpannerQueryOptions we'll add the setting to allow it . @balopat

Reply:
I like the QueryOptions idea! Then we can add that into the ref-doc as well to the partial read sections - and update the Read vs Query matrix too.

Reply:
changes made